### PR TITLE
Feat/strapi plugin

### PR DESCRIPTION
**Pull request by Truemedia**
This adds support for the [Strapi CMS](https://strapi.io/).

Supports permissions (public and user specific single login), multiple entities, custom endpoint url, and query limit.

The only features I am aware of not currently implemented yet are:
- Relations
- JSON data type

Because it's still quite usable I think these features can be released at a later point.

### CODE REVIEW

**Comment by Truemedia** [comment]
Notice I had linter errors and fix them all except one (no-unpublished-require). Reading up struggling to understand how to fix this as noone has explained it in plain english.

**Comment by hjvedvik** [comment]
Thanks for the PR! I'll try to have a look at it one of the following days. You can just rename `devDependencies` to `dependencies` in `package.json` to fix that linter error :)

**Comment by Truemedia** [comment]
Cheers that's done it :D

**Comment by hjvedvik** [comment]
Great :) The `no-unpublished-require` error name can be a bit confusing. It means that the required package won't be installed in a project using this plugin. The `dev` dependencies are only installed when working in the core repository.

**Comment by Truemedia** [comment]
Found some bugs and have done another push

File: packages/source-strapi/index.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        }
+      } catch (e) {
+        console.error('Strapi authentication error: ' + e)
+      }
+
+      console.timeEnd('Authenticate Strapi user')
+    }
+
+    return Promise.all(contentTypes.map(resourceName => {
+      const typeName = pascalCase(resourceName)
```
Comment (hjvedvik) [inline packages/source-strapi/index.js]:
```suggestion
      const typeName = store.makeTypeName(resourceName)
```
The plugin should require a `typeName` option which defaults to `Strapi` instead of the hard-coded type name.

File: packages/source-strapi/index.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      var contentType = addContentType({ typeName: `Strapi${typeName}` })
+      return query({ apiURL, contentType: typeName, jwtToken, queryLimit })
+        .then(docs => docs.map(doc => {
+          const content = {}
+          const fields = {}
+          Object.entries(doc).map(([name, val]) => {
+            if (val instanceof Object) {
+              val = JSON.stringify(val) // TODO: Look into making this easier to access
```
Comment (hjvedvik) [inline packages/source-strapi/index.js]: Why do you stringify the value? Could the objects be processed to make them cleaner instead?

File: packages/source-strapi/helpers/query.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+const clean = require('./clean')
+const fetch = require('node-fetch')
```
Comment (hjvedvik) [inline packages/source-strapi/helpers/query.js]: Would it be possible to use `axios` here too?

**Comment by Truemedia** [comment]
@hjvedvik I'll have a look at your suggestions over weekend and update the pull request, thanks

**Comment by Truemedia** [comment]
@hjvedvik Updated and ready for review :)

File: packages/source-strapi/index.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      } catch (e) {
+        console.error('Strapi authentication error: ' + e)
+      }
+
+      console.timeEnd('Authenticate Strapi user')
+    }
+
+    return Promise.all(contentTypes.map(resourceName => {
+      const typeName = makeTypeName(resourceName)
+      var contentType = addContentType({ typeName })
```
Comment (hjvedvik) [inline packages/source-strapi/index.js]:
```suggestion
      const contentType = addContentType({ typeName })
```

File: packages/source-strapi/package.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{
+  "version": "0.1.0",
```
Comment (hjvedvik) [inline packages/source-strapi/package.json]:
```suggestion
  "version": "0.0.0",
```

**Comment by Truemedia** [comment]
@hjvedvik I will leave these changes to @violentmagician so that we are not repeating the same code and he has a chance to review what I have so far and merge in his own changes. If nothing is committed I will pick this up in a week or two.

**Comment by violentmagician** [comment]
Sorry I have been busy with stuff. I will pick this back up next week and add support for graphql queries on strapi end.

**Comment by hjvedvik** [comment]
@violentmagician We are going to make a GraphQL source plugin sometime soon. Do you think it would be better to just use a plugin like that with the Strapi GraphQL APIs?

**Comment by Truemedia** [comment]
@hjvedvik I think with strapi you have to install a graphql plugin seperately, which isn't difficult and is maintained by the official strapi team but just means one extra step.

Might be best to support both but be configured by default to use the built-in rest API?

**Comment by ksnyde** [comment]
Is this still active? I'm just about to dip a toe into Gridsome and would love to pair it with Strapi

**Comment by Truemedia** [comment]
@ksnyde I'm not personally working on it anymore but willing to support the code that is there so far since I authored it. Feel free to give it a spin

**Comment by selfagency** [comment]
[i got sick of waiting](https://www.npmjs.com/package/gridsome-source-strapi)

**Comment by therenney** [comment]
Any reasons or blockers why this isn't merged already? @hjvedvik @Truemedia

**Comment by Truemedia** [comment]
I will be picking this up after my holiday regardless because I want to flesh it out with the two absent features mentioned above (Relations and JSON data type), that I now need myself for a project.

If this package is published before then I will push these 'bonus' features as a new branch,  elsewise help with the boxing off of this branch.

**Comment by hjvedvik** [comment]
Sounds great @Truemedia :) I did some updates to reflect changes that have happened in core since this PR was opened. I'm about to merge branches for the 0.7 release of Gridsome and might publish this plugin with it.

**Comment by Truemedia** [comment]
Feel free to publish whenever your ready, glad to see it can be used either way 😁

**Comment by hjvedvik** [comment]
@Truemedia Just published this source plugin :) It will require Gridsome v0.7 to work.

**Comment by egrechko** [comment]
Do you know if relationships are ready to be used for real projects? 

I really excited to develop something with strapi and gridsome. 

I have tested the relationship field and looks like I am able to query the correct fields with graphql.

**Comment by StefanoFrontini** [comment]
Thx  @Truemedia for developing this plugin. 
Are you still thinking of adding support for JSON Data Type and Strapi dynamic zones?
Thanks

**Comment by Truemedia** [comment]
@egrechko @StefanoFrontini I am going look at deeper integration but will be a few weeks away before I can make a start.

**Comment by a-kriya** [comment]
@Truemedia Perhaps https://www.npmjs.com/package/gridsome-source-strapi can now be unpublished since this has been merged and made available as @gridsome/source-strapi.

**Comment by selfagency** [comment]
I went ahead and unpublished it.

**Comment by tanc** [comment]
@selfagency also the docs here could do with updating: https://gridsome.org/plugins/gridsome-source-strapi

**Comment by selfagency** [comment]
@tanc that should be removed from the docs. it's automatically included based on npm package names, so i assume it's on gridsome to rebuild their docs and it should disappear.
